sentry/src/main/java/io/sentry/SentryTransaction.java

```diff
@@ -0,0 +1,218 @@
+package io.sentry;
+
+import io.sentry.protocol.Contexts;
+import io.sentry.protocol.SentryId;
+import io.sentry.util.Objects;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Date;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class SentryTransaction extends SentryBaseEvent implements ISpan {
+  /** The transaction name. */
+  private @Nullable String transaction;
+
+  /** The moment in time when span was started. */
+  private final @NotNull Date startTimestamp;
+
+  /** The moment in time when span has ended. */
+  private @Nullable Date timestamp;
+
+  /** A list of spans within this transaction. Can be empty. */
+  private final @NotNull List<Span> spans = new ArrayList<>();
```

Comment:
this list is not thread-safe, wondering if this would race or a transaction is always single thread because of thread locals?

Reply:
We cannot guarantee that `Transaction` will be used in single thread. What do you think is a better fit here - `CopyyOnRewrite` or `Synchronized` variant of the `ArrayList`?

Reply:
I'd prefer CopyOnWriteArrayList, @bruno-garcia usually has strong feelings about thread-safe APIs, lets ask him as well.

Reply:
The worry isnt' much about guarantee of an instance not being used in multiple threads. If we document Transaction is not thread-safe and therefore should not be used concurrently, that'd be OK.

The issue we have the `Scope` must be thread-safe. In backend Java when a new thread is spawned, it access the main hub to clone its state. That happens while that main hub might be getting modified.

On Android it's even more relevant given we have a single static, mutable, Hub. There we don't allow for `push` and `pop` scope though (they are no-op).

We need to discuss the right data type here for collections to be thread-safe as in not throwing in a `foreach` because they are modified etc, but also how can we make the APIs atomic. Those that need to be.

I wonder how this works in Python, /cc @untitaker

Reply:
Appending a span to the list is thread-safe in Python. When iterating over the spans we make sure not to trigger any of those exceptions as we don't pop or replace items in the list, ever.

Python's scopes are somewhat thread-safe but we try to make sure each concurrency unit basically gets its own scope.

Whether you want to support concurrency within a transaction right now is up to you but it's a valid usecase.